### PR TITLE
JSON Enpoints

### PR DESCRIPTION
**Pull request by Scalvando**
Adding in a way to communicate with the API via JSON.

This is a rough draft of what the API will look like and how it will behave.

### CODE REVIEW

File: control_server.go [hunk 1]
```diff
@@ -180,18 +185,22 @@ func (this *ControlServer) HandleUnpause(w http.ResponseWriter, r *http.Request)
...
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
 
-	if vars["type"] == "automatic" {
+	if cutover == "automatic" {
 		this.F.AutomaticCutover = true
-	} else if vars["type"] == "manual" {
+	} else if cutover == "manual" {
```
Comment (shuhaowu) [inline control_server.go:196]: lol at the `"manual"`.. I definitely wasn't careful when I wrote this code originally with regards to how to name different things. We should definitely refactor the cutover process to make things more clear (cc @floriecai )

File: control_server.go [hunk 2]
```diff
@@ -210,7 +219,13 @@ func (this *ControlServer) HandleVerify(w http.ResponseWriter, r *http.Request)
...
 	}
 
-	http.Redirect(w, r, "/", http.StatusSeeOther)
+	returnSuccess(w, r)
+}
+
+func (this *ControlServer) HandleProgress(w http.ResponseWriter, r *http.Request) {
+	status := this.fetchStatus()
+	w.Header().Set("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(status)
```
Comment (shuhaowu) [inline control_server.go:228]: Is there no error for this?

File: control_server.go [hunk 3]
```diff
@@ -86,6 +87,10 @@ func (this *ControlServer) Initialize() (err error) {
 
 	this.router = mux.NewRouter()
 	this.router.HandleFunc("/", this.HandleIndex).Methods("GET")
+	this.router.HandleFunc("/api/progress", this.HandleProgress).Headers("Content-Type", "application/json").Methods("GET")
+	this.router.HandleFunc("/api/cutover", this.HandleCutover).Headers("Content-Type", "application/json").Methods("POST")
+	this.router.HandleFunc("/api/verify", this.HandleVerify).Headers("Content-Type", "application/json").Methods("POST")
+	this.router.HandleFunc("/api/custom_script", this.HandleScript).Headers("Content-Type", "application/json").Methods("POST")
```
Comment (shuhaowu) [inline control_server.go:93]: I see the previous routes are `/api/actions/cutover`, worth it to unify the routes?

**Review by shuhaowu — APPROVED** [review]
Generally looks good, some minor nitpicks

File: control_server.go (hunk 3 above)
Comment (Scalvando) [reply control_server.go:93]: I think so

File: control_server.go (hunk 2 above)
Comment (Scalvando) [reply control_server.go:228]: It can return an error, at this point it will have to be logging I think

**Review by floriecai — APPROVED** [review]

File: control_server.go [hunk 4]
```diff
@@ -210,7 +216,16 @@ func (this *ControlServer) HandleVerify(w http.ResponseWriter, r *http.Request)
...
 	}
 
-	http.Redirect(w, r, "/", http.StatusSeeOther)
+	returnSuccess(w, r)
+}
+
+func (this *ControlServer) HandleStatus(w http.ResponseWriter, r *http.Request) {
+	status := this.fetchStatus()
+	w.Header().Set("Content-Type", "application/json")
+	err := json.NewEncoder(w).Encode(status.Progress)
```
Comment (Lincoln23) [inline control_server.go:225]: We want to encode the whole `status` and not only `status.Progress` right?

Comment (Scalvando) [reply control_server.go:225]: Yup, I was messing around with things and forgot to revert this change

**Review by Lincoln23 — APPROVED** [review]

**Pull request merged**